### PR TITLE
Add REST-API endpoints for property sheet schema storage.

### PR DESCRIPTION
**Pull request by deiferni**
With this PR we add REST-API endpoints to make use of the property sheet storage that was added in https://github.com/4teamwork/opengever.core/pull/6795.

The initial implementation only allows `POST` and `DELETE` for a `Manager`. `GET` is available for anyone with the `View` permission. Incremental edits for schemas are not yet supported, e.g. field removal or reordering.

Property sheets support being represented as JSON Schema as of this Pull-Request. The schema of a single sheet can be retrieved, including the schemas in the `@schema` endpoints for existing types will follow in future Pull-Request when we implement property sheet/custom field support for certain content-types.

We add a new service endpoints `@propertysheets` which is available on the plone-site only. It can be used to manage and list property sheets.

- Retrieve a list of all available sheets or the JSON-schema of a single sheet:
  `GET /@propertysheets`
  `GET /@propertysheets/<property_sheet_name>`

- Create or replace existing property sheets:
  `POST /@propertysheets/<property_sheet_name>`

- Delete existing property sheets:
  `DELETE /@propertysheets/<property_sheet_name>`

We also add the vocabulary `opengever.propertysheets.PropertySheetAssignmentsVocabulary` to validate property sheet assignments. Currently property sheets can only be assigned to potential `document_type` attribute values . We do not actually link the sheet to the `document_type` in any way, this will be implemented in the future.

_The API endpoints are not defined in `opengever.api` but in the `opengever.propertysheets.api` package. I would propose we start adding specific API classes to their corresponding sub-package and start breaking up the large package `opengever.api`._

Jira: https://4teamwork.atlassian.net/browse/CA-1279

## Checklist (Must have)

_Everything has to be done/checked. Checked but not present means the author deemed it unnecessary._

- [x] Changelog entry
- [x] Link to issue (Jira or GitHub) and backlink in issue (Jira)

## Checklist (if applicable)

_Only applicable should be left and checked._

- API change:
  - [x] Documentation is updated

### CODE REVIEW

File: opengever/propertysheets/tests/test_schema_definition_post.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+                headers=self.api_headers,
+            )
+
+        self.assertDictContainsSubset(
+            {
+                u"message": u"The name 'in-val.id' is invalid.",
+                "type": "BadRequest",
+            },
+            browser.json,
+        )
```
Comment (njohner) [inline opengever/propertysheets/tests/test_schema_definition_post.py:242]: I'd add a test here that the the storage is still empty

File: opengever/propertysheets/tests/test_schema_definition_post.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+                headers=self.api_headers,
+            )
+
+        self.assertDictContainsSubset(
+            {
+                u"message": u"Missing or invalid field definitions.",
+                "type": "BadRequest",
+            },
+            browser.json,
+        )
```
Comment (njohner) [inline opengever/propertysheets/tests/test_schema_definition_post.py:265]: Same here, test that storage remained empty

File: opengever/propertysheets/tests/test_schema_definition_post.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+                method="POST",
+                data=json.dumps(data),
+                headers=self.api_headers,
+            )
+        self.assertDictContainsSubset(
+            {
+                "type": "BadRequest",
+            },
+            browser.json,
+        )
```
Comment (njohner) [inline opengever/propertysheets/tests/test_schema_definition_post.py:284]: same same

File: opengever/propertysheets/api/delete.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+from zope.interface import implementer
+from zope.publisher.interfaces import IPublishTraverse
+
+
+@implementer(IPublishTraverse)
+class PropertySheetsDelete(Service):
+    """
+    Delete existing property sheets.
+
+    DELETE http://localhost:8080/fd/@propertysheets/question HTTP/1.1
```
Comment (njohner) [inline opengever/propertysheets/api/delete.py:15]: http://localhost:8080/fd/@propertysheets/question -> http://localhost:8080/fd/@propertysheets/sheet-id

File: opengever/propertysheets/tests/test_schema_definition_delete.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
+from ftw.testbrowser import browsing
+from opengever.propertysheets.definition import PropertySheetSchemaDefinition
+from opengever.propertysheets.storage import PropertySheetSchemaStorage
+from opengever.testing import IntegrationTestCase
+
+
+class TestSchemaDefinitionDelete(IntegrationTestCase):
```
Comment (njohner) [inline opengever/propertysheets/tests/test_schema_definition_delete.py:7]: newline

File: opengever/propertysheets/api/get.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    Return a list of all registered sheets or the schema for a single sheet.
+
+    A list of all sheets can be obtained with no additional path segments:
+
+    GET http://localhost:8080/fd/@propertysheets HTTP/1.1
+
+    The schema for a single sheet can be obtained by using the name of the
+    sheet in the request path:
+
+    GET http://localhost:8080/fd/@propertysheets/question HTTP/1.1
```
Comment (njohner) [inline opengever/propertysheets/api/get.py:20]: same here: `question` -> `sheet-id`

File: opengever/propertysheets/api/get.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        raise BadRequest(u"Must supply either zero or one parameters.")
+
+    def reply_list_all_sheets(self):
+        storage = PropertySheetSchemaStorage()
+
+        base_url = "{}/@propertysheets".format(self.context.absolute_url())
+        result = {"@id": base_url, "propertysheets": []}
+        for schema_definition in storage.list():
+            sheet_definition = {"@id": "{}/{}".format(base_url, schema_definition.name)}
+            result["propertysheets"].append(sheet_definition)
```
Comment (njohner) [inline opengever/propertysheets/api/get.py:47]: I'd change that to `result["items"]`. We might have to batch this endpoint at some point in the future, so I think using `items` here is more consistent and future proof.

File: opengever/propertysheets/assignment.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+        document_types_vocabulary = vocabulary_factory(context)
+
+        assignment_terms = []
+        for term in document_types_vocabulary:
+            name = u"IDocumentMetadata.document_type.{}".format(
+                term.value
+            )
+            assignment_terms.append(SimpleTerm(name))
+
+        return SimpleVocabulary(assignment_terms)
```
Comment (njohner) [inline opengever/propertysheets/assignment.py:25]: So that should be fine because `opengever.document.document_types` is basically an elephant vocabulary right? We never delete existing document types, but make them not choosable.

File: docs/public/dev-manual/api/propertysheets.rst [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+Property Sheets (Benutzerdefinierte Felder)
+===========================================
+
+Mittels Property Sheets ist es möglich benutzerdefinierte Schemata mit einem
+oder mehreren Feldern zu definieren damit zusätzliche Properties in GEVER
+strukturiert erfasst werden können.
+
+Im Moment ist das Definieren der zusätzlichen Property Sheets einem ``Manager``
+vorbehalten. Hierzu steht der Service-Endpoint ``@propertysheets`` sheets zur
```
Comment (njohner) [inline docs/public/dev-manual/api/propertysheets.rst:11]: ✂️ "sheets"

File: docs/public/dev-manual/api/propertysheets.rst [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+
+Schema eines Property Sheets:
+
+.. sourcecode:: http
+
+  GET /@propertysheets/<property_sheet_name> HTTP/1.1
+
+
+Hinzufügen eines neuen Property Sheets:
```
Comment (njohner) [inline docs/public/dev-manual/api/propertysheets.rst:28]: Hinzufügen eines neuen Property Sheets oder Überschreiben eines existierenden Property Sheets?

**Review by njohner — APPROVED** [review]
Nice, LGTM!
A few more points:
- first commit message: vocabuylaries -> vocabularies and vocabylary -> vocabulary :-D
- Typo in commit message title: "Add vocabulary for propert sheet assignments." -> Add vocabulary for property sheet assignments.
- Also I wonder whether we would need to add these endpoints in the API changelog (as soon as it is merged)?

File: opengever/propertysheets/api/get.py (hunk 7 above)
Comment (deiferni) [reply opengever/propertysheets/api/get.py:47]: I'm not really expecting hundreds of items, honestly 😄  … maybe dozens. but i think `items` is a good name and i have renamed it. unfortunately we don't really do it consistently yet.

File: opengever/propertysheets/assignment.py (hunk 8 above)
Comment (deiferni) [reply opengever/propertysheets/assignment.py:25]: Yes there are two layers involved, first the vdex vocabulary which always defines all terms, second is a registry setting to control term visibility. I only use the vdex, ignoring visibility to be on the safe side.

Comment (njohner) [reply opengever/propertysheets/assignment.py:25]: Yeah that's what I'd seen, was just making sure ;-)

**Comment by deiferni** [comment]
@njohner updated with your suggestions and also added also very briefly mentioned the change in the recently introduced api changelog. IMO 🚢 .

**Comment by njohner** [comment]
Yes 🚢
